pubsub/streaming-analytics/PubSubToGCS.py

```diff
@@ -14,119 +14,126 @@
+        shard_id = str(batch[0])
+        filename = "-".join([self.output_path, window_start, window_end, shard_id])
 
-        with beam.io.gcp.gcsio.GcsIO().open(filename=filename, mode="w") as f:
-            for element in batch:
-                f.write("{}\n".format(json.dumps(element)).encode("utf-8"))
+        with io.gcsio.GcsIO().open(filename=filename, mode="w") as f:
+            for message in batch[1]:
+                f.write(f"{message[0]},{message[1]}\n".encode("utf-8"))
 
 
-def run(input_topic, output_path, window_size=1.0, pipeline_args=None):
-    # `save_main_session` is set to true because some DoFn's rely on
-    # globally imported modules.
+def run(input_topic, output_path, window_size=1.0, num_shards=1, pipeline_args=None):
+    # Set `save_main_session` to True so DoFns can access globally imported modules.
     pipeline_options = PipelineOptions(
         pipeline_args, streaming=True, save_main_session=True
     )
 
-    with beam.Pipeline(options=pipeline_options) as pipeline:
+    with Pipeline(options=pipeline_options) as pipeline:
         (
             pipeline
-            | "Read PubSub Messages"
-            >> beam.io.ReadFromPubSub(topic=input_topic)
-            | "Window into" >> GroupWindowsIntoBatches(window_size)
-            | "Write to GCS" >> beam.ParDo(WriteBatchesToGCS(output_path))
+            # Because `timestamp_attribute` is unspecified in `ReadFromPubSub`, Beam
+            # binds the publish time returned by the Pub/Sub server for each message
+            # to the element. Element timestamp is accessible via `DoFn.TimestampParam`.
+            | "Read from Pub/Sub" >> io.ReadFromPubSub(topic=input_topic)
+            | "Window into" >> GroupMessagesByFixedWindows(window_size, num_shards)
+            | "Write to GCS" >> ParDo(WriteToGCS(output_path))
         )
 
 
-if __name__ == "__main__":  # noqa
+if __name__ == "__main__":
     logging.getLogger().setLevel(logging.INFO)
 
     parser = argparse.ArgumentParser()
     parser.add_argument(
         "--input_topic",
-        help="The Cloud Pub/Sub topic to read from.\n"
-        '"projects/<PROJECT_NAME>/topics/<TOPIC_NAME>".',
+        help="The Cloud Pub/Sub topic to read from."
+        '"projects/<PROJECT_ID>/topics/<TOPIC_ID>".',
     )
     parser.add_argument(
         "--window_size",
         type=float,
         default=1.0,
-        help="Output file's window size in number of minutes.",
+        help="Output file's window size in minutes.",
     )
     parser.add_argument(
         "--output_path",
-        help="GCS Path of the output file including filename prefix.",
+        help="Path of the output GCS file including the prefix.",
+    )
+    parser.add_argument(
+        "--num_shards",
+        type=int,
+        default=1,
+        help="Number of shards to use when writing windowed elements to GCS.",
     )
     known_args, pipeline_args = parser.parse_known_args()
 
     run(
         known_args.input_topic,
```

Comment:
Nit: these are starting to grow, it might be a good idea to start doing named arguments so that the order doesn't matter, otherwise it might be error-prone introducing more arguments.

```py
run(
  input_topic=known_args.input_topic,
  output_path=known_args.output_path,
  ...
)
```

Reply:
Ack.